src/amd64/amd64.c.inc

```diff
@@ -0,0 +1,201 @@
+
+#include <string.h>
+#include "ge25519.h"
+
+extern void choose_tp(ge25519_pniels *t, unsigned long long pos, signed long long b, const ge25519_pniels *base_multiples);
+
+/* return 0 on success, -1 otherwise. Taken from
+ed25519/amd64-51-30k/ge25519_unpackneg.c - the negation is removed. */
+static int unpack_vartime(ge25519_p3 *r, const unsigned char p[32])
+{
+  fe25519 t, chk, num, den, den2, den4, den6;
+  unsigned char par = p[31] >> 7;
+
+  fe25519_setint(&r->z,1);
+  fe25519_unpack(&r->y, p); 
+  fe25519_square(&num, &r->y); /* x = y^2 */
+  fe25519_mul(&den, &num, &ecd); /* den = dy^2 */
+  fe25519_sub(&num, &num, &r->z); /* x = y^2-1 */
+  fe25519_add(&den, &r->z, &den); /* den = dy^2+1 */
+
+  /* Computation of sqrt(num/den)
+     1.: computation of num^((p-5)/8)*den^((7p-35)/8) = (num*den^7)^((p-5)/8)
+  */
+  fe25519_square(&den2, &den);
+  fe25519_square(&den4, &den2);
+  fe25519_mul(&den6, &den4, &den2);
+  fe25519_mul(&t, &den6, &num);
+  fe25519_mul(&t, &t, &den);
+
+  fe25519_pow2523(&t, &t);
+  /* 2. computation of r->x = t * num * den^3
+  */
+  fe25519_mul(&t, &t, &num);
+  fe25519_mul(&t, &t, &den);
+  fe25519_mul(&t, &t, &den);
+  fe25519_mul(&r->x, &t, &den);
+
+  /* 3. Check whether sqrt computation gave correct result, multiply by sqrt(-1) if not:
+  */
+  fe25519_square(&chk, &r->x);
+  fe25519_mul(&chk, &chk, &den);
+  if (!fe25519_iseq_vartime(&chk, &num))
+    fe25519_mul(&r->x, &r->x, &sqrtm1);
+
+  /* 4. Now we have one of the two square roots, except if input was not a square
+  */
+  fe25519_square(&chk, &r->x);
+  fe25519_mul(&chk, &chk, &den);
+  if (!fe25519_iseq_vartime(&chk, &num))
+    return -1;
+
+  /* 5. Choose the desired square root according to parity:
+  */
+  if(fe25519_getparity(&r->x) == (1-par)) // only change from original function was `!=` -> `==`
+    fe25519_neg(&r->x, &r->x);
+
+  fe25519_mul(&r->t, &r->x, &r->y);
+  return 0;
+}
+
+static void p3_to_pniels(ge25519_pniels* out, ge25519_p3 const* src)
+{
+  fe25519_sub(&out->ysubx, &src->y, &src->x);
+  fe25519_add(&out->xaddy, &src->x, &src->y);
+  fe25519_mul(&out->t2d, &src->t, &ec2d);
+  out->z = src->z;
+}
+
+static void negate(ge25519* out)
+{
+  fe25519_neg(&out->x, &out->x);
+  fe25519_neg(&out->t, &out->t);
+}
+
+// similar to loops in existing implementation, but uses dynamic table instead of fixed `G`.
+static void monero_scalarmult(ge25519_p3* r, ge25519_pniels const* base, char const* sec)
+{
+  signed char b[64];
+  ge25519_pniels t;
+  ge25519_p1p1 tp1p1;
+
+  sc25519 s;
+  memcpy(s.v, sec, sizeof(s));
+  sc25519_window4(b, &s);
+
+  // set neutral
+  fe25519_setint(&r->x, 0);
+  fe25519_setint(&r->y, 1);
+  fe25519_setint(&r->t, 0);
+  fe25519_setint(&r->z, 1);
+  // end set neutral
+
+  for (int i = 63; /* break below*/ ; --i)
+  {
+    choose_tp(&t, (unsigned long long) 0, (signed long long) b[i], base);
+    ge25519_pnielsadd_p1p1(&tp1p1, r, &t);
+    ge25519_p1p1_to_p2((ge25519_p2*)r, &tp1p1);
+
+    if (i == 0) break;
```

Comment:
I would expect the implementation to be faster if the loop were unrolled to eliminate the `break` at 0 (i.e. decrease the iteration count by one and execute the three needed statements after the loop with a fixed index). Compilers usually don't figure out this variant of unrolling for you, though it's easy to check the generated assembly. I also wonder whether an explicit alignment for `b` would help; it should fit in a 64-byte cache line on Intel and AMD processors, but only if the compiler picks a suitable alignment.

Reply:
Gcc 9.2 with `-ofast` didn't unroll the loop either way. This isn't exactly a "hot loop" since many expensive functions are called from within it.